Use dynamicDowncast<T> even more in the DOM

cdumez opened the pull request:
#### 3d8458014db3038a39b984ce5763cc66c79af7fc
<pre>
Use dynamicDowncast&lt;T&gt; even more in the DOM
<a href="https://bugs.webkit.org/show_bug.cgi?id=264715">https://bugs.webkit.org/show_bug.cgi?id=264715</a>

Reviewed by Brent Fulgham.

Use dynamicDowncast&lt;T&gt; even more in the DOM instead of is&lt;T&gt;() + downcast&lt;T&gt;().
It is less error-prone and often results in more concise code. I am also hoping
to have downcast&lt;&gt;() do a type check on release builds in the future. It is
currently too expensive to do so but it may become affordable if we use
dynamicDowncast&lt;T&gt;() instead when possible.

* Source/WebCore/dom/Node.cpp:
(WebCore::Node::dumpStatistics):
(WebCore::Node::childNodes):
(WebCore::Node::insertBefore):
(WebCore::Node::replaceChild):
(WebCore::Node::removeChild):
(WebCore::Node::appendChild):
(WebCore::Node::normalize):
(WebCore::Node::computeEditabilityWithStyle const):
(WebCore::Node::renderRect):
(WebCore::Node::pseudoAwarePreviousSibling const):
(WebCore::Node::pseudoAwareNextSibling const):
(WebCore::Node::pseudoAwareFirstChild const):
(WebCore::Node::pseudoAwareLastChild const):
(WebCore::Node::isClosedShadowHidden const):
(WebCore::Node::parentInComposedTree const):
(WebCore::Node::parentElementInComposedTree const):
(WebCore::Node::parentOrShadowHostElement const):
(WebCore::Node::shadowIncludingRoot const):
(WebCore::Node::rootEditableElement const):
(WebCore::appendAttributeDesc):
(WebCore::Node::showNodePathForThis const):
(WebCore::traverseSubtreeToUpdateTreeScope):
(WebCore::Node::moveNodeToNewDocument):
(WebCore::Node::dispatchDOMActivateEvent):
(WebCore::Node::defaultEventHandler):
(WebCore::Node::willRespondToMouseMoveEvents const):
(WebCore::Node::willRespondToMouseClickEventsWithEditability const):
(WebCore::Node::removedLastRef):
* Source/WebCore/dom/Node.h:
(WebCore::EventTarget::ref):
(WebCore::EventTarget::deref):
* Source/WebCore/dom/NodeTraversal.cpp:
(WebCore::NodeTraversal::previousIncludingPseudo):
(WebCore::NodeTraversal::nextIncludingPseudo):
(WebCore::NodeTraversal::nextIncludingPseudoSkippingChildren):
* Source/WebCore/dom/Position.cpp:
(WebCore::hasInlineRun):
(WebCore::Position::containerOrParentElement const):
(WebCore::Position::anchorElementAncestor const):
(WebCore::Position::upstream const):
(WebCore::Position::downstream const):
(WebCore::Position::hasRenderedNonAnonymousDescendantsWithHeight):
(WebCore::Position::isCandidate const):
(WebCore::Position::isRenderedCharacter const):
(WebCore::Position::rendersInDifferentPosition const):
(WebCore::Position::leadingWhitespacePosition const):
(WebCore::isNonTextLeafChild):
(WebCore::searchAheadForBetterMatch):
(WebCore::Position::inlineBoxAndOffset const):
* Source/WebCore/dom/Position.h:
(WebCore::offsetIsBeforeLastNodeOffset):
* Source/WebCore/dom/PositionIterator.cpp:
(WebCore::PositionIterator::isCandidate const):
* Source/WebCore/dom/Range.cpp:
(WebCore::processContentsBetweenOffsets):
(WebCore::Range::insertNode):
(WebCore::Range::toString const):
(WebCore::Range::createContextualFragment):
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::requestModuleScript):
(WebCore::downcastScriptElement):
* Source/WebCore/dom/ScriptExecutionContext.cpp:
(WebCore::ScriptExecutionContext::reportUnhandledPromiseRejection):
(WebCore::ScriptExecutionContext::ensureRejectedPromiseTrackerSlow):
(WebCore::ScriptExecutionContext::globalObject const):
(WebCore::ScriptExecutionContext::allowsMediaDevices const):
(WebCore::ScriptExecutionContext::serviceWorkerContainer):
(WebCore::ScriptExecutionContext::ensureServiceWorkerContainer):
(WebCore::ScriptExecutionContext::postTaskToResponsibleDocument):
* Source/WebCore/dom/SelectorQuery.cpp:
(WebCore::elementsForLocalName):
* Source/WebCore/dom/ShadowRoot.h:
(WebCore::Node::isUserAgentShadowRoot const):
(WebCore::Node::parentOrShadowHostNode const):
* Source/WebCore/dom/SimulatedClick.cpp:
* Source/WebCore/dom/SlotAssignment.cpp:
(WebCore::nextSlotElementSkippingSubtree):
* Source/WebCore/dom/StyledElement.cpp:
(WebCore::StyledElement::ensureMutableInlineStyle):
(WebCore::StyledElement::setInlineStyleFromString):
* Source/WebCore/dom/Text.cpp:
(WebCore::earliestLogicallyAdjacentTextNode):
(WebCore::latestLogicallyAdjacentTextNode):
(WebCore::isSVGShadowText):
(WebCore::isSVGText):
* Source/WebCore/dom/TextNodeTraversal.h:
(WebCore::TextNodeTraversal::firstTextChildTemplate):
(WebCore::TextNodeTraversal::firstTextWithinTemplate):
(WebCore::TextNodeTraversal::traverseNextTextTemplate):
(WebCore::TextNodeTraversal::nextSibling):
* Source/WebCore/dom/TreeScope.cpp:
(WebCore::TreeScope::elementsFromPoint):
(WebCore::focusedFrameOwnerElement):
* Source/WebCore/dom/TreeScopeOrderedMap.cpp:
(WebCore::TreeScopeOrderedMap::getElementByMapName const):
(WebCore::TreeScopeOrderedMap::getElementByUsemap const):
* Source/WebCore/dom/mac/ImageControlsMac.cpp:
(WebCore::ImageControlsMac::createImageControls):
(WebCore::ImageControlsMac::imageFromImageElementNode):
(WebCore::ImageControlsMac::handleEvent):
(WebCore::ImageControlsMac::destroyImageControls):

Canonical link: <a href="https://commits.webkit.org/270654@main">https://commits.webkit.org/270654@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5849a92ddb4dc520b4d1cc798ceab72ca1605087

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25958 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4569 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27239 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28058 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23773 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/26276 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6331 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1997 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23852 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26210 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3450 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22364 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28638 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3070 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23317 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29382 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23685 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23694 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27259 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3103 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1311 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4493 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3560 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3340 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3421 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -2521,27 +2523,31 @@ void Node::defaultEventHandler(Event& event)
-
-            if (renderer) {
+            RenderBox* renderBox = nullptr;
+            for (RenderObject* renderer = this->renderer(); renderer; renderer = renderer->parent()) {
```
It seems like there should be an iterator for this (like parentsOfType) that handles the traversal of parents. But maybe that only exists for Nodes.

brentfulgham commented on Source/WebCore/dom/Range.cpp:
```diff
@@ -510,10 +510,10 @@ static ExceptionOr<RefPtr<Node>> processContentsBetweenOffsets(Range::ActionType
     case Node::PROCESSING_INSTRUCTION_NODE: {
         auto& instruction = downcast<ProcessingInstruction>(*container);
-        endOffset = std::min(endOffset, downcast<ProcessingInstruction>(*container).data().length());
+        endOffset = std::min(endOffset, instruction.data().length());
```
Ha! A wasteful double-downcast!

brentfulgham reviewed (approved):
Nice clean-up. r=me

rniwa commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -2521,27 +2523,31 @@ void Node::defaultEventHandler(Event& event)
-                renderer = renderer->parent();
-
-            if (renderer) {
+            RenderBox* renderBox = nullptr;
```
Hm... I think the old code was easier to read.

cdumez replied:
I will look into it. If you have a solution you prefer that relies on dynamic cast, I'd be happy to take it.
Personally, I felt that the new code was as clear, just less concise.

cdumez replied:
Making a simplification proposal via https://github.com/WebKit/WebKit/pull/20432

cdumez committed:
Use dynamicDowncast<T> even more in the DOM
https://bugs.webkit.org/show_bug.cgi?id=264715

Reviewed by Brent Fulgham.

Use dynamicDowncast<T> even more in the DOM instead of is<T>() + downcast<T>().
It is less error-prone and often results in more concise code. I am also hoping
to have downcast<>() do a type check on release builds in the future. It is
currently too expensive to do so but it may become affordable if we use
dynamicDowncast<T>() instead when possible.

* Source/WebCore/dom/Node.cpp:
(WebCore::Node::dumpStatistics):
(WebCore::Node::childNodes):
(WebCore::Node::insertBefore):
(WebCore::Node::replaceChild):
(WebCore::Node::removeChild):
(WebCore::Node::appendChild):
(WebCore::Node::normalize):
(WebCore::Node::computeEditabilityWithStyle const):
(WebCore::Node::renderRect):
(WebCore::Node::pseudoAwarePreviousSibling const):
(WebCore::Node::pseudoAwareNextSibling const):
(WebCore::Node::pseudoAwareFirstChild const):
(WebCore::Node::pseudoAwareLastChild const):
(WebCore::Node::isClosedShadowHidden const):
(WebCore::Node::parentInComposedTree const):
(WebCore::Node::parentElementInComposedTree const):
(WebCore::Node::parentOrShadowHostElement const):
(WebCore::Node::shadowIncludingRoot const):
(WebCore::Node::rootEditableElement const):
(WebCore::appendAttributeDesc):
(WebCore::Node::showNodePathForThis const):
(WebCore::traverseSubtreeToUpdateTreeScope):
(WebCore::Node::moveNodeToNewDocument):
(WebCore::Node::dispatchDOMActivateEvent):
(WebCore::Node::defaultEventHandler):
(WebCore::Node::willRespondToMouseMoveEvents const):
(WebCore::Node::willRespondToMouseClickEventsWithEditability const):
(WebCore::Node::removedLastRef):
* Source/WebCore/dom/Node.h:
(WebCore::EventTarget::ref):
(WebCore::EventTarget::deref):
* Source/WebCore/dom/NodeTraversal.cpp:
(WebCore::NodeTraversal::previousIncludingPseudo):
(WebCore::NodeTraversal::nextIncludingPseudo):
(WebCore::NodeTraversal::nextIncludingPseudoSkippingChildren):
* Source/WebCore/dom/Position.cpp:
(WebCore::hasInlineRun):
(WebCore::Position::containerOrParentElement const):
(WebCore::Position::anchorElementAncestor const):
(WebCore::Position::upstream const):
(WebCore::Position::downstream const):
(WebCore::Position::hasRenderedNonAnonymousDescendantsWithHeight):
(WebCore::Position::isCandidate const):
(WebCore::Position::isRenderedCharacter const):
(WebCore::Position::rendersInDifferentPosition const):
(WebCore::Position::leadingWhitespacePosition const):
(WebCore::isNonTextLeafChild):
(WebCore::searchAheadForBetterMatch):
(WebCore::Position::inlineBoxAndOffset const):
* Source/WebCore/dom/Position.h:
(WebCore::offsetIsBeforeLastNodeOffset):
* Source/WebCore/dom/PositionIterator.cpp:
(WebCore::PositionIterator::isCandidate const):
* Source/WebCore/dom/Range.cpp:
(WebCore::processContentsBetweenOffsets):
(WebCore::Range::insertNode):
(WebCore::Range::toString const):
(WebCore::Range::createContextualFragment):
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::requestModuleScript):
(WebCore::downcastScriptElement):
* Source/WebCore/dom/ScriptExecutionContext.cpp:
(WebCore::ScriptExecutionContext::reportUnhandledPromiseRejection):
(WebCore::ScriptExecutionContext::ensureRejectedPromiseTrackerSlow):
(WebCore::ScriptExecutionContext::globalObject const):
(WebCore::ScriptExecutionContext::allowsMediaDevices const):
(WebCore::ScriptExecutionContext::serviceWorkerContainer):
(WebCore::ScriptExecutionContext::ensureServiceWorkerContainer):
(WebCore::ScriptExecutionContext::postTaskToResponsibleDocument):
* Source/WebCore/dom/SelectorQuery.cpp:
(WebCore::elementsForLocalName):
* Source/WebCore/dom/ShadowRoot.h:
(WebCore::Node::isUserAgentShadowRoot const):
(WebCore::Node::parentOrShadowHostNode const):
* Source/WebCore/dom/SimulatedClick.cpp:
* Source/WebCore/dom/SlotAssignment.cpp:
(WebCore::nextSlotElementSkippingSubtree):
* Source/WebCore/dom/StyledElement.cpp:
(WebCore::StyledElement::ensureMutableInlineStyle):
(WebCore::StyledElement::setInlineStyleFromString):
* Source/WebCore/dom/Text.cpp:
(WebCore::earliestLogicallyAdjacentTextNode):
(WebCore::latestLogicallyAdjacentTextNode):
(WebCore::isSVGShadowText):
(WebCore::isSVGText):
* Source/WebCore/dom/TextNodeTraversal.h:
(WebCore::TextNodeTraversal::firstTextChildTemplate):
(WebCore::TextNodeTraversal::firstTextWithinTemplate):
(WebCore::TextNodeTraversal::traverseNextTextTemplate):
(WebCore::TextNodeTraversal::nextSibling):
* Source/WebCore/dom/TreeScope.cpp:
(WebCore::TreeScope::elementsFromPoint):
(WebCore::focusedFrameOwnerElement):
* Source/WebCore/dom/TreeScopeOrderedMap.cpp:
(WebCore::TreeScopeOrderedMap::getElementByMapName const):
(WebCore::TreeScopeOrderedMap::getElementByUsemap const):
* Source/WebCore/dom/mac/ImageControlsMac.cpp:
(WebCore::ImageControlsMac::createImageControls):
(WebCore::ImageControlsMac::imageFromImageElementNode):
(WebCore::ImageControlsMac::handleEvent):
(WebCore::ImageControlsMac::destroyImageControls):

Canonical link: https://commits.webkit.org/270654@main
